Added with_* and validate functions

cs-victor-nascimento opened the pull request:
This is to help on the ongoing effort of refactoring the code base. I know this is incomplete and feel free to deny this PR, but this implements partially (without tests, documentation and type specification) what is being discussed in #62 on top of your initial changes. 

What we have here is:

- all default with_* functions.
- verifying a token
- some refactoring of where the code belongs to (moved signing and verifying to signer)

The intention is to help us move forward in the discussion. With this PR you can do this in iex:

```elixir
import Joken

token 
|> with_sub("I am the subject") 
|> with_validation(:sub, %(&1 == "I am the subject"))
|> with_HS512("my ultra secret secret") # convenience function
|> sign # also convenience. generates the token
|> verify # generates the claims map from a token
```

This still signs/verifies as soon as you call those functions and it returns the struct always. Following some other discussions here I changed errors to error (because the user will not have to pattern match on a list of errors) and assumed some more defaults (mainly Poison). I renamed payload to claims as that is more explicit to me, but feel free to revert that. 

I haven't implemented tests, documentation and specs because I feel like this is a good point to discuss a bit more.

cs-victor-nascimento commented:
One thing I was thinking about is that we may want to have a `sign/1` and `verify/1` instead of `with_HS256/2`.

I mean this:
```elixir
# hs256 is only a convenience
token
|> signer(hs256("mySecret")) # this adds the signer to the struct. 
|> sign
|> verify
```
... is the same as:
```elixir
# hs256 is only a convenience
token
|> sign(hs256("mySecret"))
|> verify(hs256("mySecret"))
```

The difference being portability. We could make `sign/2`override default signer. This will make Plug's configuration easier.

Just some food for thought.

bryanjos commented:
I like what you have here. having `claims` instead of `payload` does make more sense. I was also thinking that using the signer struct as a parameter would be too low-level in most situations so I like your ideas about creating convenience functions around it. A definite plus if it makes the Plug configuration easier. If it's ready I can merge it.

cs-victor-nascimento commented:
I think this work is to be done in multiple steps. You can merge this one and I will open another PR to adapt this for convenience functions. What you say?

cs-victor-nascimento commented:
Disconsider my last comment. I've thought about it and it is better to make a PR with all that set. I will update this PR today.

bryanjos commented:
Sounds good

cs-victor-nascimento commented:
Okay. I added the signer configuration as stated in the comments and fixed Signer's convenience functions to generate a proper Signer. 

This also has:
- Assumes Poison as default JSON module (and passes this configuration down to erlang-jose)
- Assumes more defaults in `token()` function
- Adds some basic tests for signing/verifying using all HS*** (we need to much better here I know... but it is something)
- Added some documentation but not nearly enough for it all

This still has to figure out some configuration for `Joken.Plug` as I realized we can't have it as initially thought. We can't simply call `plug Joken.Plug, token_config: joken() |> with_signer(hs256("secret"))` or pass it through a module attribute. I will think about it.

cs-victor-nascimento commented:
Completed documentation and added specs for Joken, Joken.Token and Joken.Signer.

I would like some feedback on the following:

- Should we keep the old API along side the new one? The only issue with maintaining both is moduledocs which will have to cite both. 
- This is focused on the main API. There are some options currently available (skip_claims, skip_verifying and claim validation options) that are not handled in this API. How do you think we can solve those? 
- The validation here assumes that if there is no validation for a claim, it will be accepted. I am really not sure if this is good and if I should invert that. 

Regards!

bryanjos commented:
@cs-victor-nascimento good work!

* We can probably remove the old API I think.
* I think skip claims is the easiest one since the token is just passed to functions. If someone wanted to skip certain claims in certain situations they have more flexibility to do so now. skip_verifying I don't have any ideas on yet. I'll have to think about it.
* Accepting a claim with no validation I think is fine. It's what I would expect at least. But if something about that doesn't seem right, we can change it.

cs-victor-nascimento commented:
Do you want to make it all on this PR?

If so I will change its title.

Cheers,
Em 19/08/2015 12:09 AM, "Bryan Joseph" <notifications@github.com> escreveu:

> @cs-victor-nascimento <https://github.com/cs-victor-nascimento> good work!
>
>    - We can probably remove the old API I think.
>    - I think skip claims is the easiest one since the token is just
>    passed to functions. If someone wanted to skip certain claims in certain
>    situations they have more flexibility to do so now. skip_verifying I don't
>    have any ideas on yet. I'll have to think about it.
>    - Accepting a claim with no validation I think is fine. It's what I
>    would expect at least. But if something about that doesn't seem right, we
>    can change it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bryanjos/joken/pull/64#issuecomment-132429521>.
>

bryanjos commented:
No, we can do it separately. I'm going to merge this now
